Handle EXCLUDE constraints properly on distributed tables

robin900 opened the pull request:
...just as we handle UNIQUE or PRIMARY KEY constraints. If an EXCLUDE constraint includes
the distribution column, we verify that the exclusion operator for the column implements equality. Otherwise we raise an error and forbid the exclusion constraint.

Fixes #748 and #778.

onderkalaci commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -185,24 +185,25 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
-		/* only check unique indexes */
-		if (indexInfo->ii_Unique == false)
+		/* only check unique indexes and exclusion constraints. */
+		if (indexInfo->ii_Unique == false && indexInfo->ii_ExclusionOps == NULL)
```
Hey @robin900 thanks for your contribution!

>  However, it will be more laborious to check that the exclusion operator for the distribution column is indeed '=': there are lots of equality operators in pg_operator, varying by argument data type, which would be valid for Citus to accept, and some of which have an oprname != '='.

In #748 you mentioned the above, which I agree.  

But the good news is that in Citus, we already have a function called [OperatorImplementsEquality(opno](https://github.com/citusdata/citus/blob/6333f9ba6f2c6af605ca1528764283e44e7efb15/src/backend/distributed/planner/multi_logical_planner.c#L2046)) which is exactly what you need. Given an operator number, returns true if it is an equality operator.

 I think it'd be convinient if we do something like the following in the following loop:
```C

		for (attributeIndex = 0; attributeIndex < attributeCount; attributeIndex++)
		{
			AttrNumber attributeNumber = attributeNumberArray[attributeIndex];
			if (indexInfo->ii_Unique && distributionColumn->varattno == attributeNumber)
			{
				hasDistributionColumn = true;
				break;
			}
    
			if (indexInfo->ii_ExclusionOps != NULL && 
			    distributionColumn->varattno == attributeNumber && 
			    OperatorImplementsEquality((indexInfo->ii_ExclusionOps [attributeIndex]))
			{
				hasDistributionColumn = true;
				break;
			}
		}
```
(Well, the above is just a psudo-code, but hope that it shows what I wanted to explain. We could probably write the code with nested `if` statements where the top level is`if distributionColumn->varattno == attributeNumber`)

onderkalaci commented on src/test/regress/expected/multi_create_table.out:
```diff
@@ -122,15 +122,24 @@ CREATE TABLE primary_key_on_non_part_col
-DETAIL:  Distributed relations cannot have UNIQUE constraints or PRIMARY KEYs that do not include the partition column.
+SELECT master_create_distributed_table('unique_const_on_non_part_col', 'partition_col', 'hash');
+ERROR:  cannot distribute relation: "unique_const_on_non_part_col"
+DETAIL:  Distributed relations cannot have UNIQUE constraints, EXCLUDE constraints, or PRIMARY KEYs that do not include the partition column.
```
If we're to implement my suggestion above, this error message becomes obsolete for `EXCLUDE` constraints. We should probably have two different error messages, where in `EXCLUDE` constraints' error message we mention that partition column should be specified with an equality operator.

robin900 commented:
@onderkalaci I've added the OperatorImplementsEquality() call. I've changed the error message to mention that an EXCLUDE constraint must include the partition column with an equality operator. The message you mention does not seem to me to be obsolete; and I'm not feeling the desire to make more complex error reporting logic just to return a customized errdetail() for the unique constraint case, and the exclusion constraint case, etc.

Tell me if the latest commit's changes are sufficient for you...

robin900 commented:
Argh, while rebasing my feature branch to citus/master, I accidentally added some now-removed test files from long ago in master. Force push of corrected code complete.

onderkalaci commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -235,8 +237,17 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 		{
-			AttrNumber attributeNumber = attributeNumberArray[attributeIndex];
-			if (distributionColumn->varattno == attributeNumber)
+			if (distributionColumn->varattno != attributeNumberArray[attributeIndex])
```
I think having an explicit variable assignement is more convinient wrt Citus conventions. In this case, it'd be better if we keep `AttrNumber attributeNumber` and continue to use it.

onderkalaci commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -235,8 +237,17 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
+			{
+				continue;
+			}
+			if (indexInfo->ii_Unique)
```
We prefer to have a new line between if statement such as:

```C
			AttrNumber attributeNumber = attributeNumberArray[attributeIndex];
			if (distributionColumn->varattno != attributeNumber)
			{
				continue;
			}

			if (indexInfo->ii_Unique)
			{
				hasDistributionColumn = true;
				break;
			}

			if (indexInfo->ii_ExclusionOps != NULL &&
			    OperatorImplementsEquality(indexInfo->ii_ExclusionOps[attributeIndex]))
			{
				hasDistributionColumn = true;
				break;
			}

```

onderkalaci commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -209,24 +210,25 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 									 distributedRelationName),
-							  errdetail("Unique constraints and primary keys on "
-										"append-partitioned tables cannot be enforced."),
+							  errdetail(
```
We have 90 characters hard limit. Error detail seems to pass the limit. Also, we do not start the detail with an empty line. Could you please check other error message formatting in the same file?

robin900 replied:
The 90-character limit is for source code, or for the error detail message itself?

onderkalaci commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -249,13 +260,15 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 		index_close(indexDesc, NoLock);
 	}
 
+
```
Unnecessary new line here.

onderkalaci commented on src/test/regress/output/multi_subquery.source:
```diff
@@ -861,7 +861,7 @@ DEBUG:  predicate pruning for shardId 270011
 DEBUG:  predicate pruning for shardId 270012
 DEBUG:  predicate pruning for shardId 270011
 DEBUG:  predicate pruning for shardId 270012
-DEBUG:  building index "pg_toast_17247_index" on table "pg_toast_17247"
```
Well, this test still seems to fail. Probably you forgot to update after adding the test with `exclusion_const_wrong_operator ` table?

onderkalaci commented:
Hey @robin900 , I think the PR is almost ready to check-in. 
I added a few coding convention related comments. Also, `multi_subquery` test is failing which I think very easy to fix and added a comment there as well. 

After addressing the minor comments, could you also rebase to the master branch so that we'd be ready to check-in?

Thanks!

robin900 commented:
@onderkalaci just force-pushed a new feature branch that resolves the code formatting issues. (I interpreted your "90 character limit" to mean source code lines must be <= 90 characters wide. Also, the multi_subquery test failure will happen only if the tmp_check data directory hasn't been removed since the last test run. In a clean test environment, the test passes. (The difference is in the pg toast table names' embedding a different, larger counter value if the PG instance isn't built from scratch.)

onderkalaci commented:
This was quick @robin900  :):)

> The difference is in the pg toast table names' embedding a different, larger counter value if the PG instance isn't built from scratch.

Why this doesn't happen on the master branch and happens here? Any ideas?

onderkalaci commented:
> Why this doesn't happen on the master branch and happens here? Any ideas?
 Show all checks

Oops, that happens on the master as well, sorry.

onderkalaci commented:
@robin900 I now realized that Citus does not push down EXCLUDE constraints. See the below:

```SQL
CREATE TABLE exclusion_const_on_part_col   ( 
partition_col integer,                                                                                                                                                          other_col integer,                                                                                                                                                              CONSTRAINT exclusion_const_on_part_col_excl EXCLUDE (partition_col WITH =)  );

SELECT master_create_distributed_table('exclusion_const_on_part_col', 'partition_col', 'hash');
SELECT master_create_worker_shards('exclusion_const_on_part_col', 2, 1);

\d exclusion_const_on_part_col

Table "public.exclusion_const_on_part_col"
    Column     |  Type   | Modifiers 
---------------+---------+-----------
 partition_col | integer | 
 other_col     | integer | 
Indexes:
    "exclusion_const_on_part_col_excl" EXCLUDE USING btree (partition_col WITH =)

\c - - - 9700

\d exclusion_const_on_part_col*

Table "public.exclusion_const_on_part_col_102008"
    Column     |  Type   | Modifiers 
---------------+---------+-----------
 partition_col | integer | 
 other_col     | integer | 
Indexes:
    "exclusion_const_on_part_col_excl_102008" btree (partition_col)

Index "public.exclusion_const_on_part_col_excl_102008"
    Column     |  Type   |  Definition   
---------------+---------+---------------
 partition_col | integer | partition_col
btree, for table "public.exclusion_const_on_part_col_102008"

```

I don't really think that it is a blocker to check-in this PR. It probably requires to open a new PR for that. But, still I'll need to check with @sumedhpathak or @marcocitus .

In one place you @robin900 mentioned that you use EXCLUSION constraints. Could you please check that the constraints are actually there?

robin900 commented:
@onderkalaci While evaluating Citus, I did not check whether the exclusion constraints were propagated to the shard tables. After reviewing master/master_node_protocol.c, it seems clear that they were not propagated correctly. pg_get_indexdef_string() doesn't handle exclusion ops and thus the resulting index definition never happens. A one-line change to master_node_protocol.c lets pg_get_constraintdef_string() do the work and all seems well. Amended commit force-pushed to feature branch, and all tests passed.

robin900 commented:
@onderkalaci Of course, the regression suite does not verify that the index/constraint defs propagate to shard tables in any of the test cases in multi_create_table.c. I'm not up to adding the sufficient test scaffolding to check this. Given that it's a one line change to master_node_protocol.c, and the reliability of pg_get_constraintdef_string(), it might be sufficient for you just to run your interactive test again, see that the constraint propagated, and we're all good for now.

robin900 commented:
@onderkalaci Ok, I added checks in the regression suite that valid exclusion constraints propagate, by doing INSERTs that would violate the exclusion constraints. The new checks in the regression tests run successfully now in any timezone (the Travis server's UTC server timezone differs from my MBP's). This PR should be ready to go.

onderkalaci commented on src/test/regress/expected/multi_create_table.out:
```diff
@@ -173,14 +182,89 @@ SELECT master_create_distributed_table('unique_const_on_two_columns', 'partition
+ 
+(1 row)
+
+INSERT INTO exclusion_const_on_non_part_col (partition_col, other_col) VALUES (1,1);
```
I guess these `INSERT`s should go to `exclusion_const_on_part_col` and we'd see errors?

onderkalaci commented on src/backend/distributed/master/master_node_protocol.c:
```diff
@@ -648,7 +648,7 @@ GetTableDDLEvents(Oid relationId)
 			isConstraint = true;
 		}
-		else if (indexForm->indisunique)
+		else if (indexForm->indisunique || indexForm->indisexclusion)
```
I think this change is not sufficient. We also need to add shardId to the exclude constraint index on the workers to prevent creating the same named exclusion indexes when replication factor > 2. See the example below:
```SQL

CREATE TABLE exclusion_const_overlaps_operator
(
	partition_col tsrange,
	other_col tsrange,
	CONSTRAINT exclusion_const_overlaps_operator_excl EXCLUDE USING gist (other_col WITH &&, partition_col WITH =)
);

SELECT master_create_distributed_table('exclusion_const_overlaps_operator', 'partition_col', 'hash');

SELECT master_create_worker_shards('exclusion_const_overlaps_operator', '2', '2');

WARNING:  relation "exclusion_const_overlaps_operator_excl" already exists
CONTEXT:  while executing command on localhost:9701
WARNING:  could not create shard on "localhost:9701"
ERROR:  could only create 1 of 2 of required shard replicas
```
Which probably ties into `RelayEventExtendNames()` function.

(We should also consider dropping the table as we consider creating shards)

aamederen replied:
Would be nice to mention exclusion alongside uniqueness in the comment section above the if statements

onderkalaci commented on src/test/regress/regression.out:
```diff
@@ -0,0 +1,11 @@
+test multi_extension          ... ok
```
Ops, I guess we should not be adding `regression.out` here.

onderkalaci commented:
@robin900 We're getting close :) 

I added a few more comments. Especially one of them is critical, creating >2 shards on a single worker errors out given that we do not append shard ids to the constraint names.

Could you also address that, and possibly change shard count & replication factor one of the tests so that it covers that condition as well?

Thanks!

metdos commented:
I've fixed the problem related with multi_subquery on the master branch. A rebase should fix the problem here also.

robin900 commented:
Rebased to get Metin's changes. So three problems:

1) Named constraints of any kind (unique, exclusion) can't be propagated to shards. I recommend making a separate issue in github for that.

2) Using unnamed constraints (just base UNIQUE or EXCLUDE), I cannot test the enforcement of the constraint (and thus its existence on the shards) because the error message will mention the name implicitly given to the constraint, and that will vary.

3) Even unnamed, exclusion constraints have a problem because PG implicitly names them ${relation_name}_excl", and so there's a name collision.

I'd appreciate any advice on how to proceed. I don't know where the best place would be to add the logic to add shardIds to DDL...

robin900 commented:
@onderkalaci I've factored the constraint-related regression tests into their own file in the suite. I have added sufficient tests for PK, UNIQUE, and EXCLUDE constraints both named and unnamed. Now I am wrestling with the remaining problem, which is that indexes/constraints on a distributed table may be too long to append the shard ID for it.

I made issue #781 about the name length problem, since it affects even distributed table names. In this PR's branch, I'm adding all the fixes for too-long names, and with that, support for valid unique and exclusion constraints will be complete, afaik.

robin900 commented:
@onderkalaci I've separated the name-length issue completely from this PR, mainly by keeping table and constraint names short enough in the regression tests.  robin900/citus has a [conform-to-DISTNAMEDATALEN branch](https://github.com/robin900/citus/tree/conform-to-DISTNAMEDATALEN) that begins to address the name-length issue.

This PR now, as far as I can tell, propagates DDL correctly for unique and exclusion constraints, supports both named and unnamed constraints and appends shardId properly to both named and unnamed (implicitly named) constraints. Tests pass. Let me know if you see anything else...

robin900 commented:
@onderkalaci @sumedhpathak I am hopeful that this can be placed in the 5.3 milestone...

onderkalaci commented:
> @onderkalaci @sumedhpathak I am hopeful that this can be placed in the 5.3 milestone...

@robin900 We also want to check this in. @aamederen will continue the review.

aamederen commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -236,7 +239,20 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
+			}
+
+			if (indexInfo->ii_ExclusionOps != NULL &&
+				OperatorImplementsEquality(indexInfo->ii_ExclusionOps[attributeIndex]))
```
Instead of having two `if` statements with the same body, could we merge them with some helper variables to something like this:

```
bool uniqueConstraint = indexInfo->ii_Unique;
bool exclusionConstraintWithEquality = indexInfo->ii_ExclusionOps != NULL && OperatorImplementsEquality(indexInfo->ii_ExclusionOps[attributeIndex]);

if (uniqueConstraint || exclusionConstraintWithEquality)
{
    hasDistributionColumn = true;
    break;
}

```

aamederen commented on src/backend/distributed/relay/relay_event_utility.c:
```diff
@@ -497,7 +498,8 @@ TypeDropIndexConstraint(const AlterTableCmd *command,
 			if (constraintForm->contype == CONSTRAINT_PRIMARY ||
-				constraintForm->contype == CONSTRAINT_UNIQUE)
+				constraintForm->contype == CONSTRAINT_UNIQUE ||
+				constraintForm->contype == CONSTRAINT_EXCLUSION)
```
We have this kind of check twice in the codebase. I think we could extract them in a function and avoid the duplicate logic.

aamederen replied:
Mention exclusion constraint alongside uniqueness and primary key in the comments above this section, starting with "At this stage..."

robin900 replied:
@aamederen The data structures, and even the enums, differ. One is from the parse tree, with a `ConstrType` enum. The other is after a catalog lookup, checking the constraint type as a char. May we skip this refactoring in the PR?

aamederen replied:
@robin900 I see, let it be then 👍

aamederen commented on src/test/regress/sql/multi_create_table_constraints.sql:
```diff
@@ -0,0 +1,208 @@
+	partition_col integer UNIQUE,
+	other_col integer
+);
+SELECT master_create_distributed_table('uq_range_tables', 'partition_col', 'range');
```
It can be beneficial to include some more tests about constraints with IMMUTABLE functions and partial constraints.

Adding a CHECK constraint test would also be nice but it is not directly related to this PR.

aamederen replied:
Also a comprehensive test showing that we support multiple EXCLUDE constraints would be nice.

robin900 commented:
@aamederen I've made the requested changes to code comments and error/notice messages. I've added test coverage of a CHECK constraint and a CHECK constraint using an IMMUTABLE function (abs). 

I have not added any coverage of partial index constraints, because [Postgres does not allow partial unique indexes _as constraints_, only as indexes](http://stackoverflow.com/a/16236566). I suggest instead adding such tests to `multi_index_statements.sql`.

ozgune commented:
@robin900 -- quick note. @sumedhpathak is OOTO for a few days. I'll mention this pull request to him later this week when he's back.

aamederen commented on src/backend/distributed/commands/create_distributed_table.c:
```diff
@@ -203,29 +204,33 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
 		AttrNumber *attributeNumberArray = NULL;
 		bool hasDistributionColumn = false;
+		bool uniqueConstraint = false;
+		bool exclusionConstraintWithEquality = false;
```
These variables could be moved inside the for block where they have been used.

aamederen commented:
Onee more subtle change and we can :shipit: 

@ozgune, @sumedhpathak asked me to do the testing/review for this PR. I think we can merge it after the last changes.

robin900 commented:
@aamederen moved inside the block.

aamederen commented:
@robin900 Looks Great! Thanks for the contribution.
